### ISSUE TITLE
Add `SnippetEnter` and `SnippetLeave` events

### ISSUE THREAD

**Issue by echasnovski** [neovim/neovim#26449]
### Problem

Currently there is no proper enough way for users (plugins) to act when snippet session is started or ended. This is useful, for example, in the following cases:
- Set mapping to be active only when snippet session is active.
- Create autocommand to be active only when snippet session is active. Like to stop snippet session when exiting to `Normal` mode by using `ModeChanged` event (created on `SnippetEnter`, deleted on `SnippetLeave`).

Both these use cases *can* be worked around by having `if vim.snippet.active()` check inside of them, but it seems very wasteful to have mapping/autocommand all the time when it is only needed during active snippet session.

### Expected behavior

`SnippetEnter` and `SnippetLeave` `User` events which trigger when snippet session starts and ends. Ideally, with some snippet parsing data parsed in `data` of autocommand's callback.

**Comment by justinmk** [comment]
> Set mapping to be active only when snippet session is active

Mappings during a snippet seems overwrought, but otoh 'keymap' depends on mappings, for example.

A new event seems clumsy for this, though, since it sounds like you actually want a "mode": https://github.com/neovim/neovim/issues/992

It's common for snippet plugins (e.g. UltiSnips) to use `:help Select-mode` for snippets. Are we using that?

**Comment by clason** [comment]
Let's not be premature here. Events and autocommands are _the_ way Neovim communicates state, and we do want people to be able to hook into and customize snippet expansion. While the sample use case in the issue alone is not necessarily a compelling argument, I would like to collect other use cases that can be handled with this; if there's enough of them, this would be worth it -- especially since it'd be straightforward to add and document (@mariasolos?), compared to the alternative of exposing currently private functions.

> It's common for snippet plugins (e.g. UltiSnips) to use :help Select-mode for snippets. Are we using that?

Yes, but only for placeholders. It doesn't apply to tabstops without "default text", so this is irrelevant to this issue.

**Comment by echasnovski** [comment]
> A new event seems clumsy for this, though, since it sounds like you actually want a "mode": #992

What I actually want (right now) is a clean way to implement "stop snippet session after exit to Normal mode" behavior.

**Comment by mfussenegger** [comment]
> What I actually want (right now) is a clean way to implement "stop snippet session after exit to Normal mode" behavior.


To add for completeness sake, the current workarounds for that concrete use-case:

a) Override all keys that switch to normal mode, to call snippet.exit, e.g.:

```
keymap.set({'i', 's'}, '<ESC>', function()
  if vim.snippet then
    vim.snippet.exit()
  end
  return '<ESC>'
end, { expr = true })
```


b) Call `vim.snippet.exit()` in a `ModeChanged` event.


---


Downsides:
- Cost of `vim.snippet.exit()` always firing (Maybe ns or microsecond range?)
- Cost of the `ModeChanged` event always firing if using `b)`
- Upside: Shorter than setting up a `SnippetEnter` & `SnippetLeave` that manage `ModeChanged`

**Comment by MariaSolOs** [comment]
I'm in favour of having snippet events and moving the cancellation logic responsibility to the user. We've had issues with cancellation before (e.g. https://github.com/neovim/neovim/pull/25835).

>> I would like to collect other use cases that can be handled with this; if there's enough of them, this would be worth it -- especially since it'd be straightforward to add and document (@MariaSolOs?), 

Given the current "experimental" state of the snippet API, I much rather prefer this approach than exposing a changing interface.

**Comment by echasnovski** [comment]
> > What I actually want (right now) is a clean way to implement "stop snippet session after exit to Normal mode" behavior.
> 
> To add for completeness sake, the current workarounds for that concrete use-case:
> 
> a) Override all keys that switch to normal mode, to call snippet.exit, e.g.:
> 
> ```
> keymap.set({'i', 's'}, '<ESC>', function()
>   if vim.snippet then
>     vim.snippet.exit()
>   end
>   return '<ESC>'
> end, { expr = true })
> ```
> 
> b) Call `vim.snippet.exit()` in a `ModeChanged` event.

After actually trying to use `ModeChanged` event, it doesn't really work in all cases. It can get triggered when navigating through tabstops non empty text (either previously entered text and going back to it or a tabstop with placeholder). This seems to be because tabstop selection itself exits to Normal mode: [here](https://github.com/neovim/neovim/blob/b2d471ab337e56f660eb7c89ae24f447f7b7a165/runtime/lua/vim/snippet.lua#L279) and [here](https://github.com/neovim/neovim/blob/b2d471ab337e56f660eb7c89ae24f447f7b7a165/runtime/lua/vim/snippet.lua#L289).

So under current implementation I don't "stop snippet session if user exited to Normal mode" is possible with autocommand. Using a global remap for `<Esc>` is not really an option in the case of a general plugin (which relies on `vim.snippet` functionality). Making that remap being local and only during snippet session is rather painful but acceptable.

**Comment by mfussenegger** [comment]
> I'm in favour of having snippet events and moving the cancellation logic responsibility to the user. We've had issues with cancellation before 

If there is no cancellation by default it also means users cannot use snippets at all without defining their own logic, as they would forever stay in snippet mode.
I'd even tend towards going the opposite direction: Set more defaults for snippets (like keymaps). In part because not having any defaults means we cannot have snippet expansion enabled for completion by default either. 


Neovim lists strong defaults as a feature/goal and I think we should try and achieve that, as long as it's not too disruptive and controversial. I'm not yet convinced that snippet cancellation is _that_ opinionated, and that we can't come up with a default behavior that will work good enough for most users.

**Comment by MariaSolOs** [comment]
> If there is no cancellation by default it also means users cannot use snippets at all without defining their own logic, as they would forever stay in snippet mode.

Right, I wasn't suggesting on moving the responsibility _entirely_ ;)

> I'd even tend towards going the opposite direction: Set more defaults for snippets (like keymaps). In part because not having any defaults means we cannot have snippet expansion enabled for completion by default either. That means we can't set the LSP capability, which means users who do want to use it have to do more configuration work.

I like this and I think we should keep this goal. However maybe we should make clear what the purpose of those defaults are, and what level of customization (if at all) we're willing to accommodate.

**Comment by justinmk** [comment]
sketch of "user defined modes": https://github.com/neovim/neovim/issues/992#issuecomment-2577779050
